runtime/lua/vim/diagnostic.lua

```diff
@@ -1584,6 +1591,26 @@ local function render_virtual_text(namespace, bufnr, diagnostics, opts)
   end
 end
 
+--- Displays the given diagnostics as virtual text.
+---
+--- @param namespace integer Diagnostic namespace ID
+--- @param bufnr? integer Buffer handle, or 0 for current (default: 0)
+--- @param diagnostics vim.Diagnostic[] Diagnostics to display
+--- @param opts? vim.diagnostic.Opts.VirtualText
+function M.render_virtual_text(namespace, bufnr, diagnostics, opts)
```

Comment:
I made this a separate function to keep the `diagnostics` argument a list of diagnostics, and not a map from line to line diagnostics.